Sprint 12

Dimaulta opened the pull request:
Задачи по итогам 12-го спринта

Dimaulta committed:
Инициализация 12-го спринта

Dimaulta committed:
Статус приоржения до:
"Самостоятельная задача. Реализуйте в проекте постраничную загрузку данных для ленты фотографий"

Dimaulta committed:
Структуры для декодирования JSON (PhotoResult, UrlsResult)
Структура Photo для UI
Класс ImagesListService с полной функциональностью
Интеграция с UI и вызов fetchPhotosNextPage
Обработка обновления данных через нотификации

Dimaulta committed:
Устраняю проблему с Git

Dimaulta committed:
Стабильная работа приложения image Feed
Реализовал бесконечную прокрутку ленты настоящих фотографий из сервиса Unsplash
Добавил загрузку unsplash по макету
Добавил в SingleImageViewController загрузку реальных фотографий вместо заглушки
Использовал Kingfisher для загрузки и кеширование изображения

Dimaulta committed:
Добавил функционал лайков и их грамотное отображение в ленте фотографий

Dimaulta committed:
Доработал SingleImage

Dimaulta committed:
Добавил LogoutService с возвратом на экран входа

Dimaulta committed:
Добавил алерт который появляется при нажатии на кнопку выхода из приложения
Проверил статус приложения по чек-листу 12-го спринта

Helvetii951 commented on ImageFeed/AppDelegate.swift:
```diff
@@ -6,15 +6,16 @@
     func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
-        // Override point for customization after application launch.
-        // sleep(1)
+   
+        ProgressHUD.animationType = .activityIndicator
+        ProgressHUD.colorHUD = .black
+        ProgressHUD.colorAnimation = .white
```
Лучше выносить подобные настройки в отдельную функцию. Иначе [application(_:didFinishLaunchingWithOptions:)](https://developer.apple.com/documentation/uikit/uiapplicationdelegate/1622921-application) может стать громоздкой.

А еще лучше было бы перенести в статическую функцию в обертку UIBlockingProgressHUD`.

Helvetii951 commented on ImageFeed/ImagesList/ImagesListService.swift:
```diff
@@ -0,0 +1,136 @@
+
+import UIKit
+
+struct PhotoResult: Decodable {
+    let id: String
+    let width: Int
+    let height: Int
+    let createdAt: String?
+    let description: String?
+    let likedByUser: Bool
+    let urls: UrlsResult
+    
+    enum CodingKeys: String, CodingKey {
+        case id, width, height, description, urls
+        case createdAt = "created_at"
+        case likedByUser = "liked_by_user"
+    }
+}
+
+struct UrlsResult: Decodable {
+    let thumb: String
+    let full: String
+}
+
+struct Photo {
+    let id: String
+    let size: CGSize
+    let createdAt: Date?
+    let welcomeDescription: String?
+    let thumbImageURL: String
+    let largeImageURL: String
+    let isLiked: Bool
+}
+
+final class ImagesListService {
```
Лучше создавать для каждой сущности один файл. Так навигация по проекту будет понятнее. Т.е. стоит создать для `UserResult` отдельный файл.

Helvetii951 commented on ImageFeed/ImagesList/ImagesListService.swift:
```diff
@@ -0,0 +1,136 @@
+                    let decoder = JSONDecoder()
+                    let photoResults = try decoder.decode([PhotoResult].self, from: data)
+                    let newPhotos = photoResults.map { result in
+                        Photo(
+                            id: result.id,
+                            size: CGSize(width: result.width, height: result.height),
+                            createdAt: ISO8601DateFormatter().date(from: result.createdAt ?? ""),
+                            welcomeDescription: result.description,
+                            thumbImageURL: result.urls.thumb,
+                            largeImageURL: result.urls.full,
+                            isLiked: result.likedByUser
+                        )
```
Было бы классно создать инициализатор для `Photo`
```swift
init(from : PhotoResult) {  
  // здесь перекидываем свойcтва PhotoResponseResult
  return createdPhoto
}
```
Так мы сможем значительно сок ратить код сервиса. Плюс будет удобно использовать с функциональными операторами [map(_:), flatMap(_:) и compactMap(_:)](https://swiftblog.org/funkcii-map-flatmap-i-compactmap/).

Helvetii951 commented on ImageFeed/ImagesList/ImagesListService.swift:
```diff
@@ -0,0 +1,136 @@
+        isLoading = true
+        
+        let nextPage = (lastLoadedPage ?? 0) + 1
+        let urlString = "https://api.unsplash.com/photos?page=\(nextPage)&per_page=\(perPage)"
```
Очень желательно брать `URL`ы из файла с константами. Так не нужно будет менять код сервиса, если что-то изменится в бэкэнде.

Helvetii951 commented on ImageFeed/ImagesList/ImagesListViewController.swift:
```diff
@@ -47,52 +79,114 @@ final class ImagesListViewController: UIViewController {
+                if let visibleCell = self.tableView.cellForRow(at: indexPath) as? ImagesListCell {
+                    visibleCell.setIsLiked(newPhoto.isLiked)
+                }
+                
+                UIBlockingProgressHUD.dismiss()
+            case .failure:
+                UIBlockingProgressHUD.dismiss()
+                // TODO: Показать ошибку с использованием UIAlertController
+                print("Failed to change like status")
```
`UIBlockingProgressHUD.dismiss()` можно вынести на уровень `switch`.

Helvetii951 commented on ImageFeed/ImagesList/ImagesListViewController.swift:
```diff
@@ -6,27 +6,56 @@
 //
 
 import UIKit
+import Kingfisher
 
 final class ImagesListViewController: UIViewController {
     private let showSingleImageSegueIdentifier = "ShowSingleImage"
```
Можно убрать `import Kingfisher`. В классе не используется.

Helvetii951 commented on ImageFeed/ImagesList/ImagesListViewController.swift:
```diff
@@ -47,52 +79,114 @@ final class ImagesListViewController: UIViewController {
+                .transition(.fade(0.2))
+            ]
+        ) { [weak cell] result in
+            guard let cell = cell else { return }
+            switch result {
+            case .success:
+                cell.cellImage.contentMode = .scaleAspectFill
+            case .failure:
+                cell.cellImage.contentMode = .center
+            }
```
Не всегда необходимо распаковывать `self`. Тут можно использовать [Optional Chaing](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/optionalchaining/#:~:text=Optional%20chaining%20is%20a%20process,in%20the%20chain%20is%20nil):
```
cell?.cellImage.contentMode = .scaleAspectFill
```

Helvetii951 commented on ImageFeed/ImagesList/ImagesListViewController.swift:
```diff
@@ -47,52 +79,114 @@ final class ImagesListViewController: UIViewController {
+        let photo = photos[indexPath.row]
+        guard let imageURL = URL(string: photo.largeImageURL) else { return }
+        
+        cell.cellImage.backgroundColor = UIColor(named: "YP Gray")
+        cell.cellImage.contentMode = .center
+        cell.cellImage.kf.indicatorType = .activity
```
Рекомендую перенести настройку UI компонентов внутрь ячейки, сделать компоненты приватными. 
```swift
func configureCell(with dateString: String, другие параметры)
```
Если изменится что-то в UI ячейки, нам не нужно будет менять контроллер.

Dimaulta committed:
Форматтер создаётся один раз при инициализации класса и переиспользуется для всех преобразований.
Поменял расположение кнопок логаут-алерта местами и сделал их style: .default для соответствия макету.
